File: packages/go/hub.go
package aikit

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"time"
)

type Config struct {
	OpenAI         *OpenAIConfig
	Anthropic      *AnthropicConfig
	XAI            *XAIConfig
	Google         *GoogleConfig
	Ollama         *OllamaConfig
	HTTPClient     *http.Client
	RegistryTTL    time.Duration
	Adapters       map[Provider]ProviderAdapter
	AdapterFactory AdapterFactory
}

type OpenAIConfig struct {
	APIKey              string
	APIKeys             []string
	BaseURL             string
	Organization        string
	DefaultUseResponses bool
}

type AnthropicConfig struct {
	APIKey  string
	APIKeys []string
	BaseURL string
	Version string
}

type XAIConfig struct {
	APIKey            string
	APIKeys           []string
	BaseURL           string
	CompatibilityMode string
}

type GoogleConfig struct {
	APIKey  string
	APIKeys []string
	BaseURL string
}

type OllamaConfig struct {
	APIKey              string
	APIKeys             []string
	BaseURL             string
	DefaultUseResponses bool
}

type ProviderAdapter interface {
	ListModels(ctx context.Context) ([]ModelMetadata, error)
	Generate(ctx context.Context, in GenerateInput) (GenerateOutput, error)
	GenerateImage(ctx context.Context, in ImageGenerateInput) (ImageGenerateOutput, error)
	GenerateMesh(ctx context.Context, in MeshGenerateInput) (MeshGenerateOutput, error)
	Transcribe(ctx context.Context, in TranscribeInput) (TranscribeOutput, error)
	Stream(ctx context.Context, in GenerateInput) (<-chan StreamChunk, error)
}

type AdapterFactory func(provider Provider, entitlement *EntitlementContext) (ProviderAdapter, error)

type Kit struct {
	adapters map[Provider]ProviderAdapter
	registry *modelRegistry
	factory  AdapterFactory
	keyPools map[Provider]*keyPool
}

func New(config Config) (*Kit, error) {
	adapters := make(map[Provider]ProviderAdapter)
	keyPools := make(map[Provider]*keyPool)
	client := config.HTTPClient
	if client == nil {
		client = http.DefaultClient
	}

	for provider, adapter := range config.Adapters {
		if adapter != nil {
			adapters[provider] = adapter
		}
	}

	if config.OpenAI != nil && adapters[ProviderOpenAI] == nil {
		keys := normalizeKeys(config.OpenAI.APIKey, config.OpenAI.APIKeys)
		if len(keys) == 0 {
			return nil, fmt.Errorf("openai api key is required")
		}
		cfg := *config.OpenAI
		cfg.APIKey = keys[0]
		adapters[ProviderOpenAI] = newOpenAIAdapter(&cfg, client, ProviderOpenAI)
		keyPools[ProviderOpenAI] = newKeyPool(keys)
	}
	if config.Anthropic != nil && adapters[ProviderAnthropic] == nil {
		keys := normalizeKeys(config.Anthropic.APIKey, config.Anthropic.APIKeys)
		if len(keys) == 0 {
			return nil, fmt.Errorf("anthropic api key is required")
		}
		cfg := *config.Anthropic
		cfg.APIKey = keys[0]
		adapters[ProviderAnthropic] = newAnthropicAdapter(&cfg, client, ProviderAnthropic)
		keyPools[ProviderAnthropic] = newKeyPool(keys)
	}
	if config.XAI != nil && adapters[ProviderXAI] == nil {
		keys := normalizeKeys(config.XAI.APIKey, config.XAI.APIKeys)
		if len(keys) == 0 {
			return nil, fmt.Errorf("xai api key is required")
		}
		cfg := *config.XAI
		cfg.APIKey = keys[0]
		adapters[ProviderXAI] = newXAIAdapter(&cfg, client)
		keyPools[ProviderXAI] = newKeyPool(keys)
	}
	if config.Google != nil && adapters[ProviderGoogle] == nil {
		keys := normalizeKeys(config.Google.APIKey, config.Google.APIKeys)
		if len(keys) == 0 {
			return nil, fmt.Errorf("google api key is required")
		}
		cfg := *config.Google
		cfg.APIKey = keys[0]
		adapters[ProviderGoogle] = newGoogleAdapter(&cfg, client)
		keyPools[ProviderGoogle] = newKeyPool(keys)
	}
	if config.Ollama != nil && adapters[ProviderOllama] == nil {
		cfg := *config.Ollama
		keys := normalizeKeys(cfg.APIKey, cfg.APIKeys)
		if len(keys) > 0 {
			cfg.APIKey = keys[0]
			keyPools[ProviderOllama] = newKeyPool(keys)
		}
		adapters[ProviderOllama] = newOllamaAdapter(&cfg, client)
	}
	if len(adapters) == 0 && config.AdapterFactory == nil {
		return nil, fmt.Errorf("at least one provider config or adapter is required")
	}
	ttl := config.RegistryTTL
	if ttl == 0 {
		ttl = 30 * time.Minute
	}
	factory := config.AdapterFactory
	if factory == nil {
		factory = newAdapterFactory(config, client, adapters)
	}
	registry := newModelRegistry(adapters, ttl, factory)
	return &Kit{
		adapters: adapters,
		registry: registry,
		factory:  factory,
		keyPools: keyPools,
	}, nil
}

func (h *Kit) ListModels(ctx context.Context, opts *ListModelsOptions) ([]ModelMetadata, error) {
	models, err := h.registry.List(ctx, opts)
	if err != nil {
		return nil, err
	}
	annotateModelAvailability(models, h.adapters)
	return models, nil
}

func (h *Kit) ListModelRecords(ctx context.Context, opts *ListModelsOptions) ([]ModelRecord, error) {
	return h.registry.ListRecords(ctx, opts)
}

func annotateModelAvailability(models []ModelMetadata, adapters map[Provider]ProviderAdapter) {
	if len(models) == 0 {
		return
	}
	availableProviders := make(map[Provider]struct{}, len(adapters)+1)
	for provider := range adapters {
		availableProviders[provider] = struct{}{}
	}
	for idx, model := range models {
		_, ok := availableProviders[model.Provider]
		models[idx].Available = ok
	}
}

func (h *Kit) Generate(ctx context.Context, in GenerateInput) (GenerateOutput, error) {
	if entitlement := h.entitlementForProvider(in.Provider); entitlement != nil {
		return h.GenerateWithContext(ctx, entitlement, in)
	}
	adapter, ok := h.adapters[in.Provider]
	if !ok {
		return GenerateOutput{}, fmt.Errorf("provider %s is not configured", in.Provider)
	}
	output, err := adapter.Generate(ctx, in)
	if err != nil {
		h.registry.LearnModelUnavailable(nil, in.Provider, in.Model, err)
		return GenerateOutput{}, err
	}
	return attachCost(in, output), nil
}

func (h *Kit) GenerateWithContext(ctx context.Context, entitlement *EntitlementContext, in GenerateInput) (GenerateOutput, error) {
	adapter, err := h.factory(in.Provider, entitlement)
	if err != nil {
		return GenerateOutput{}, err
	}
	output, err := adapter.Generate(ctx, in)
	if err != nil {
		h.registry.LearnModelUnavailable(entitlement, in.Provider, in.Model, err)
	}
	return attachCost(in, output), err
}

func (h *Kit) GenerateImage(ctx context.Context, in ImageGenerateInput) (ImageGenerateOutput, error) {
	if entitlement := h.entitlementForProvider(in.Provider); entitlement != nil {
		return h.GenerateImageWithContext(ctx, entitlement, in)
	}
	adapter, ok := h.adapters[in.Provider]
	if !ok {
		return ImageGenerateOutput{}, fmt.Errorf("provider %s is not configured", in.Provider)
	}
	output, err := adapter.GenerateImage(ctx, in)
	if err != nil {
		h.registry.LearnModelUnavailable(nil, in.Provider, in.Model, err)
		return ImageGenerateOutput{}, err
	}
	return output, nil
}

func (h *Kit) GenerateImageWithContext(ctx context.Context, entitlement *EntitlementContext, in ImageGenerateInput) (ImageGenerateOutput, error) {
	adapter, err := h.factory(in.Provider, entitlement)
	if err != nil {
		return ImageGenerateOutput{}, err
	}
	output, err := adapter.GenerateImage(ctx, in)
	if err != nil {
		h.registry.LearnModelUnavailable(entitlement, in.Provider, in.Model, err)
	}
	return output, err
}

func (h *Kit) GenerateMesh(ctx context.Context, in MeshGenerateInput) (MeshGenerateOutput, error) {
	if entitlement := h.entitlementForProvider(in.Provider); entitlement != nil {
		return h.GenerateMeshWithContext(ctx, entitlement, in)
	}
	adapter, ok := h.adapters[in.Provider]
	if !ok {
		return MeshGenerateOutput{}, fmt.Errorf("provider %s is not configured", in.Provider)
	}
	output, err := adapter.GenerateMesh(ctx, in)
	if err != nil {
		h.registry.LearnModelUnavailable(nil, in.Provider, in.Model, err)
		return MeshGenerateOutput{}, err
	}
	return output, nil
}

func (h *Kit) Transcribe(ctx context.Context, in TranscribeInput) (TranscribeOutput, error) {
	if entitlement := h.entitlementForProvider(in.Provider); entitlement != nil {
		return h.TranscribeWithContext(ctx, entitlement, in)
	}
	adapter, ok := h.adapters[in.Provider]
	if !ok {
		return TranscribeOutput{}, fmt.Errorf("provider %s is not configured", in.Provider)
	}
	output, err := adapter.Transcribe(ctx, in)
	if err != nil {
		h.registry.LearnModelUnavailable(nil, in.Provider, in.Model, err)
		return TranscribeOutput{}, err
	}
	return output, nil
}

func (h *Kit) GenerateMeshWithContext(ctx context.Context, entitlement *EntitlementContext, in MeshGenerateInput) (MeshGenerateOutput, error) {
	adapter, err := h.factory(in.Provider, entitlement)
	if err != nil {
		return MeshGenerateOutput{}, err
	}
	output, err := adapter.GenerateMesh(ctx, in)
	if err != nil {
		h.registry.LearnModelUnavailable(entitlement, in.Provider, in.Model, err)
	}
	return output, err
}

func (h *Kit) TranscribeWithContext(ctx context.Context, entitlement *EntitlementContext, in TranscribeInput) (TranscribeOutput, error) {
	adapter, err := h.factory(in.Provider, entitlement)
	if err != nil {
		return TranscribeOutput{}, err
	}
	output, err := adapter.Transcribe(ctx, in)
	if err != nil {
		h.registry.LearnModelUnavailable(entitlement, in.Provider, in.Model, err)
	}
	return output, err
}

func (h *Kit) StreamGenerate(ctx context.Context, in GenerateInput) (<-chan StreamChunk, error) {
	if entitlement := h.entitlementForProvider(in.Provider); entitlement != nil {
		return h.StreamGenerateWithContext(ctx, entitlement, in)
	}
	adapter, ok := h.adapters[in.Provider]
	if !ok {
		return nil, fmt.Errorf("provider %s is not configured", in.Provider)
	}
	stream, err := adapter.Stream(ctx, in)
	if err != nil {
		return nil, err
	}
	return attachCostToStream(in.Provider, in.Model, stream), nil
}

func (h *Kit) StreamGenerateWithContext(ctx context.Context, entitlement *EntitlementContext, in GenerateInput) (<-chan StreamChunk, error) {
	adapter, err := h.factory(in.Provider, entitlement)
	if err != nil {
		return nil, err
	}
	stream, err := adapter.Stream(ctx, in)
	if err != nil {
		return nil, err
	}
	return attachCostToStream(in.Provider, in.Model, stream), nil
}

func (h *Kit) entitlementForProvider(provider Provider) *EntitlementContext {
	pool := h.keyPools[provider]
	if pool == nil {
		return nil
	}
	key := pool.Next()
	if strings.TrimSpace(key) == "" {
		return nil
	}
	return &EntitlementContext{
		Provider:          provider,
		APIKey:            key,
		APIKeyFingerprint: FingerprintAPIKey(key),
	}
}

func attachCost(in GenerateInput, output GenerateOutput) GenerateOutput {
	cost := estimateCost(in.Provider, in.Model, output.Usage)
	if cost != nil {
		output.Cost = cost
	}
	return output
}

func attachCostToStream(provider Provider, model string, stream <-chan StreamChunk) <-chan StreamChunk {
	out := make(chan StreamChunk)
	go func() {
		defer close(out)
		for chunk := range stream {
			if chunk.Type == StreamChunkMessageEnd {
				cost := estimateCost(provider, model, chunk.Usage)
				if cost != nil {
					chunk.Cost = cost
				}
			}
			out <- chunk
		}
	}()
	return out
}

func newAdapterFactory(config Config, client *http.Client, adapters map[Provider]ProviderAdapter) AdapterFactory {
	return func(provider Provider, entitlement *EntitlementContext) (ProviderAdapter, error) {
		if entitlement == nil || strings.TrimSpace(entitlement.APIKey) == "" {
			if adapter, ok := adapters[provider]; ok {
				return adapter, nil
			}
			return nil, fmt.Errorf("provider %s is not configured", provider)
		}
		apiKey := strings.TrimSpace(entitlement.APIKey)
		switch provider {
		case ProviderOpenAI:
			if config.OpenAI == nil {
				return nil, fmt.Errorf("openai config is not available")
			}
			cfg := *config.OpenAI
			cfg.APIKey = apiKey
			return newOpenAIAdapter(&cfg, client, ProviderOpenAI), nil
		case ProviderAnthropic:
			if config.Anthropic == nil {
				return nil, fmt.Errorf("anthropic config is not available")
			}
			cfg := *config.Anthropic
			cfg.APIKey = apiKey
			return newAnthropicAdapter(&cfg, client, ProviderAnthropic), nil
		case ProviderXAI:
			if config.XAI == nil {
				return nil, fmt.Errorf("xai config is not available")
			}
			cfg := *config.XAI
			cfg.APIKey = apiKey
			return newXAIAdapter(&cfg, client), nil
		case ProviderGoogle:
			if config.Google == nil {
				return nil, fmt.Errorf("google config is not available")
			}
			cfg := *config.Google
			cfg.APIKey = apiKey
			return newGoogleAdapter(&cfg, client), nil
		case ProviderOllama:
			if config.Ollama == nil {
				return nil, fmt.Errorf("ollama config is not available")
			}
			cfg := *config.Ollama
			cfg.APIKey = apiKey
			return newOllamaAdapter(&cfg, client), nil
		default:
			return nil, fmt.Errorf("provider %s is not configured", provider)
		}
	}
}

File: packages/python/src/inference_kit/hub.py
from __future__ import annotations

from dataclasses import dataclass, replace
from typing import Dict

from .catalog import CatalogAdapter, load_catalog_models
from .errors import ErrorKind, HubErrorPayload, InferenceKitError, to_hub_error
from .entitlements import fingerprint_api_key
from .pricing import estimate_cost
from .registry import ModelRegistry
from .types import (
    EntitlementContext,
    GenerateInput,
    GenerateOutput,
    ImageGenerateInput,
    ImageGenerateOutput,
    MeshGenerateInput,
    MeshGenerateOutput,
    Provider,
)
from .providers import (
    AnthropicAdapter,
    AnthropicConfig,
    GeminiAdapter,
    GeminiConfig,
    OpenAIAdapter,
    OpenAIConfig,
    XAIAdapter,
    XAIConfig,
)


@dataclass
class HubConfig:
    providers: Dict[Provider, object]
    registry_ttl_seconds: int = 1800


class _KeyPool:
    def __init__(self, keys: list[str]) -> None:
        self._keys = keys
        self._index = 0

    def next(self) -> str:
        if not self._keys:
            return ""
        key = self._keys[self._index % len(self._keys)]
        self._index = (self._index + 1) % len(self._keys)
        return key


class Hub:
    def __init__(self, config: HubConfig) -> None:
        catalog_models = load_catalog_models()
        if not config.providers and not catalog_models:
            raise InferenceKitError(
                HubErrorPayload(
                    kind=ErrorKind.VALIDATION,
                    message="At least one provider configuration is required",
                )
            )
        self._providers, self._key_pools = self._prepare_providers(config.providers)
        self._adapters = self._build_adapters(self._providers)
        if catalog_models:
            self._adapters["catalog"] = CatalogAdapter(catalog_models)
        self._registry = ModelRegistry(
            self._adapters,
            ttl_seconds=config.registry_ttl_seconds,
            adapter_factory=self._adapter_factory,
        )

    def list_models(self, providers=None, refresh=False, entitlement=None):
        return self._registry.list_models(providers, refresh, entitlement)

    def list_model_records(self, providers=None, refresh=False, entitlement=None):
        return self._registry.list_model_records(providers, refresh, entitlement)

    def generate(self, input: GenerateInput) -> GenerateOutput:
        entitlement = self._entitlement_for_provider(input.provider)
        if entitlement:
            return self.generate_with_context(entitlement, input)
        adapter = self._require_adapter(input.provider)
        try:
            output = adapter.generate(input)
            return _attach_cost(input, output)
        except Exception as err:
            hub_err = to_hub_error(err)
            self._registry.learn_model_unavailable(None, input.provider, input.model, hub_err)
            raise hub_err

    def generate_with_context(
        self, entitlement: EntitlementContext | None, input: GenerateInput
    ) -> GenerateOutput:
        adapter = self._require_adapter(input.provider, entitlement)
        try:
            output = adapter.generate(input)
            return _attach_cost(input, output)
        except Exception as err:
            hub_err = to_hub_error(err)
            self._registry.learn_model_unavailable(entitlement, input.provider, input.model, hub_err)
            raise hub_err

    def generate_image(self, input: ImageGenerateInput) -> ImageGenerateOutput:
        entitlement = self._entitlement_for_provider(input.provider)
        if entitlement:
            return self.generate_image_with_context(entitlement, input)
        adapter = self._require_adapter(input.provider)
        if not hasattr(adapter, "generate_image"):
            raise InferenceKitError(
                HubErrorPayload(
                    kind=ErrorKind.UNSUPPORTED,
                    message=f"Provider {input.provider} does not support image generation",
                )
            )
        try:
            return adapter.generate_image(input)
        except Exception as err:
            hub_err = to_hub_error(err)
            self._registry.learn_model_unavailable(None, input.provider, input.model, hub_err)
            raise hub_err

    def generate_image_with_context(
        self, entitlement: EntitlementContext | None, input: ImageGenerateInput
    ) -> ImageGenerateOutput:
        adapter = self._require_adapter(input.provider, entitlement)
        if not hasattr(adapter, "generate_image"):
            raise InferenceKitError(
                HubErrorPayload(
                    kind=ErrorKind.UNSUPPORTED,
                    message=f"Provider {input.provider} does not support image generation",
                )
            )
        try:
            return adapter.generate_image(input)
        except Exception as err:
            hub_err = to_hub_error(err)
            self._registry.learn_model_unavailable(entitlement, input.provider, input.model, hub_err)
            raise hub_err

    def generate_mesh(self, input: MeshGenerateInput) -> MeshGenerateOutput:
        entitlement = self._entitlement_for_provider(input.provider)
        if entitlement:
            return self.generate_mesh_with_context(entitlement, input)
        adapter = self._require_adapter(input.provider)
        if not hasattr(adapter, "generate_mesh"):
            raise InferenceKitError(
                HubErrorPayload(
                    kind=ErrorKind.UNSUPPORTED,
                    message=f"Provider {input.provider} does not support mesh generation",
                )
            )
        try:
            return adapter.generate_mesh(input)
        except Exception as err:
            hub_err = to_hub_error(err)
            self._registry.learn_model_unavailable(None, input.provider, input.model, hub_err)
            raise hub_err

    def generate_mesh_with_context(
        self, entitlement: EntitlementContext | None, input: MeshGenerateInput
    ) -> MeshGenerateOutput:
        adapter = self._require_adapter(input.provider, entitlement)
        if not hasattr(adapter, "generate_mesh"):
            raise InferenceKitError(
                HubErrorPayload(
                    kind=ErrorKind.UNSUPPORTED,
                    message=f"Provider {input.provider} does not support mesh generation",
                )
            )
        try:
            return adapter.generate_mesh(input)
        except Exception as err:
            hub_err = to_hub_error(err)
            self._registry.learn_model_unavailable(entitlement, input.provider, input.model, hub_err)
            raise hub_err

    def stream_generate(self, input: GenerateInput):
        entitlement = self._entitlement_for_provider(input.provider)
        if entitlement:
            return self.stream_generate_with_context(entitlement, input)
        adapter = self._require_adapter(input.provider)
        return _attach_cost_stream(adapter.stream_generate(input), input.provider, input.model)

    def stream_generate_with_context(
        self, entitlement: EntitlementContext | None, input: GenerateInput
    ):
        adapter = self._require_adapter(input.provider, entitlement)
        return _attach_cost_stream(adapter.stream_generate(input), input.provider, input.model)

    def _build_adapters(self, providers: Dict[Provider, object]):
        adapters = {}
        if "openai" in providers:
            adapters["openai"] = OpenAIAdapter(providers["openai"])
        if "anthropic" in providers:
            adapters["anthropic"] = AnthropicAdapter(providers["anthropic"])
        if "google" in providers:
            adapters["google"] = GeminiAdapter(providers["google"])
        if "xai" in providers:
            adapters["xai"] = XAIAdapter(providers["xai"])
        return adapters

    def _prepare_providers(self, providers: Dict[Provider, object]):
        key_pools: Dict[Provider, _KeyPool] = {}
        normalized: Dict[Provider, object] = {}
        for provider, cfg in providers.items():
            keys = self._collect_keys(cfg)
            if not keys:
                raise InferenceKitError(
                    HubErrorPayload(
                        kind=ErrorKind.VALIDATION,
                        message=f"Provider {provider} api key is required",
                    )
                )
            key_pools[provider] = _KeyPool(keys)
            normalized[provider] = self._with_api_key(cfg, keys[0])
        return normalized, key_pools

    def _collect_keys(self, cfg: object) -> list[str]:
        api_key = getattr(cfg, "api_key", "") or ""
        api_keys = getattr(cfg, "api_keys", None) or []
        keys = []
        seen = set()
        for raw in [api_key, *api_keys]:
            trimmed = raw.strip() if isinstance(raw, str) else ""
            if not trimmed or trimmed in seen:
                continue
            seen.add(trimmed)
            keys.append(trimmed)
        return keys

    def _with_api_key(self, cfg: object, api_key: str) -> object:
        if hasattr(cfg, "__dataclass_fields__"):
            return replace(cfg, api_key=api_key)
        if hasattr(cfg, "api_key"):
            setattr(cfg, "api_key", api_key)
            return cfg
        return cfg

    def _entitlement_for_provider(self, provider: Provider) -> EntitlementContext | None:
        pool = self._key_pools.get(provider)
        if not pool:
            return None
        api_key = pool.next()
        if not api_key:
            return None
        return EntitlementContext(
            provider=provider,
            apiKey=api_key,
            apiKeyFingerprint=fingerprint_api_key(api_key),
        )

    def _adapter_factory(
        self, provider: Provider, entitlement: EntitlementContext | None
    ):
        if provider == "catalog":
            return self._adapters.get(provider)
        if not entitlement or not entitlement.apiKey:
            return self._adapters.get(provider)
        base_config = self._providers.get(provider)
        if base_config is None:
            return None
        if provider == "openai":
            config = OpenAIConfig(
                api_key=entitlement.apiKey,
                base_url=getattr(base_config, "base_url", "https://api.openai.com"),
                organization=getattr(base_config, "organization", None),
                default_use_responses=getattr(base_config, "default_use_responses", True),
                timeout=getattr(base_config, "timeout", None),
            )
            return OpenAIAdapter(config)
        if provider == "anthropic":
            config = AnthropicConfig(
                api_key=entitlement.apiKey,
                base_url=getattr(base_config, "base_url", "https://api.anthropic.com"),
                version=getattr(base_config, "version", None) or "2023-06-01",
                timeout=getattr(base_config, "timeout", None),
            )
            return AnthropicAdapter(config)
        if provider == "google":
            config = GeminiConfig(
                api_key=entitlement.apiKey,
                base_url=getattr(base_config, "base_url", "https://generativelanguage.googleapis.com"),
                timeout=getattr(base_config, "timeout", None),
            )
            return GeminiAdapter(config)
        if provider == "xai":
            config = XAIConfig(
                api_key=entitlement.apiKey,
                base_url=getattr(base_config, "base_url", "https://api.x.ai"),
                compatibility_mode=getattr(base_config, "compatibility_mode", "openai"),
                timeout=getattr(base_config, "timeout", None),
            )
            return XAIAdapter(config)
        return None

    def _require_adapter(self, provider: Provider, entitlement: EntitlementContext | None = None):
        adapter = self._adapter_factory(provider, entitlement)
        if not adapter:
            raise InferenceKitError(
                HubErrorPayload(
                    kind=ErrorKind.VALIDATION,
                    message=f"Provider {provider} is not configured",
                )
            )
        return adapter


def _attach_cost(input: GenerateInput, output: GenerateOutput) -> GenerateOutput:
    cost = estimate_cost(input.provider, input.model, output.usage)
    if cost is None:
        return output
    return replace(output, cost=cost)


def _attach_cost_stream(stream, provider: Provider, model: str):
    for chunk in stream:
        if getattr(chunk, "type", None) == "message_end":
            cost = estimate_cost(provider, model, getattr(chunk, "usage", None))
            if cost is not None:
                yield replace(chunk, cost=cost)
                continue
        yield chunk
